metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/powerbi_api.py

```diff
@@ -0,0 +1,341 @@
+        scan_id: Optional[str] = None
+        try:
+            scan_id = self.__admin_api_resolver.create_scan_job(
+                workspace_id=workspace.id
+            )
+        except requests.exceptions.HTTPError as e:
+            if data_resolver.is_permission_error(e):
+                logger.warning(
+                    "Dataset lineage can not be ingestion because this user does not have access to the PowerBI Admin "
+                    "API. "
+                )
+                return None
+            # raise error if other than 401 or 403
+            raise e
+
+        logger.info("Waiting for scan to complete")
+        if (
+            self.__admin_api_resolver.wait_for_scan_to_complete(
+                scan_id=scan_id, timeout=self.__config.scan_timeout
+            )
+            is False
+        ):
+            raise ValueError(
+                "Workspace detail is not available. Please increase scan_timeout to wait."
+            )
+
+        # Scan is complete lets take the result
+        scan_result = self.__admin_api_resolver.get_scan_result(scan_id=scan_id)
+        pretty_json: str = json.dumps(scan_result, indent=1)
+        logger.debug(f"scan result = {pretty_json}")
+
+        return scan_result
+
+    @staticmethod
+    def _parse_endorsement(endorsements: Optional[dict]) -> List[str]:
+        if not endorsements:
+            return []
+
+        endorsement = endorsements.get("endorsement", None)
+        if not endorsement:
+            return []
+
+        return [endorsement]
+
+    def _get_workspace_datasets(self, scan_result: Optional[dict]) -> dict:
+        """
+        Filter out "dataset" from scan_result and return Dataset instance set
+        """
+        dataset_map: dict = {}
+
+        if scan_result is None:
+            return dataset_map
+
+        datasets: Optional[Any] = scan_result.get("datasets")
+        if datasets is None or len(datasets) == 0:
+            logger.warning(
+                f'Workspace {scan_result["name"]}({scan_result["id"]}) does not have datasets'
+            )
+
+            logger.info("Returning empty datasets")
+            return dataset_map
+
+        for dataset_dict in datasets:
+            dataset_instance: PowerBIDataset = self._get_resolver().get_dataset(
+                workspace_id=scan_result["id"],
+                dataset_id=dataset_dict["id"],
+            )
+
+            if self.__config.extract_endorsements_to_tags:
+                dataset_instance.tags = self._parse_endorsement(
+                    dataset_dict.get("endorsementDetails", None)
+                )
+
+            dataset_map[dataset_instance.id] = dataset_instance
+            # set dataset-name
+            dataset_name: str = (
+                dataset_instance.name
+                if dataset_instance.name is not None
+                else dataset_instance.id
+            )
+
+            for table in dataset_dict["tables"]:
```

Comment:
What happens if "tables" is Not present inside dataset_dict?

Reply:
Added default